### ISSUE TITLE
decode: speculative verify batched inside CUDA graphs (MTP K=2 to 4)

### ISSUE THREAD

**Issue by jomcgi** [jomcgi-org/freetoken-fork#9]
The CPU round trip is paid per step, not per token, so verifying K drafted tokens in one step amortises the dominant decode cost on both the 4090 (28 CPU round trips) and the Blackwell box (PCIe streaming per step). The resident MTP head already accepts 52 to 78 percent at K=1 (bench/RESULTS.md, 3 September) and lost only because the verify pass runs outside CUDA graphs (37 to 48 s versus 19 to 31 s for a 300-token answer). Work: capture the draft and verify under graphs, then batched verify at K=2 to 4 with an exact rollback of the GDN/KDA recurrent state on rejection (the part two previous attempts died on; write the rollback out explicitly before dispatching). Flag stays `--speculative-mtp off` by default. For coding workloads, prompt-lookup drafts (n-gram copies from the context, no model) are a follow-up draft source.

**Comment by jomcgi** [comment]
Code map for the spec (HEAD 30e333c), three facts that change the plan:

1. **With `--speculative-mtp on` no CUDA graph is captured at all.** `engine.py:539-551` passes an empty `cuda_graph_bs` to the `GraphRunner`, so the whole run is eager, not just the verify pass. `docs/models.md:181-183` gives the reason: the draft consumes `model.model._last_hc_hidden`, which graph replay does not expose. Making that a persistent capture buffer (like `GraphCaptureBuffer.logits`) is the prerequisite for everything else, and on its own probably accounts for most of the 37 to 48 s versus 19 to 31 s gap.
2. **The verify pass is graphable as written.** `configure_mtp_fused_step` (`spec_decode.py:40-64`) builds a fixed 2-row step, and the stateful layers run the ordinary width-1 kernels twice in a Python loop (`gdn.py:193-219`, `qsa_sparse.py:361-405`, `ple.py:3066-3080`): 36 GDN layers times 2 launches per step. Two things are not graph-stable: `build_fla_metadata` (`attention/linear.py:52-76`) allocates a fresh `arange` per call where the graph runner uses persistent `fla_cu_seqlens` and `table_idx` (`graph.py:45-59`), and the draft head (`mtp.py:186-215`) rebuilds its KV from Python lists on every call. A second graph family keyed by width (2 rows, bs 1) with persistent FLA metadata, draft head left eager, is the contained first cut.
3. **Partial acceptance at K>1 has no mechanism.** `snapshot_verify_state` / `restore_verify_state` (`spec_decode.py:67-107`) is one pre-window checkpoint of conv, recurrent, PLE slot states and the QSA pending ring; `greedy_accept_prefix` and `rollback_paged_tail` (`cache.py:638-654`) are already K-agnostic. So K=2 to 4 needs either K per-position GDN checkpoints (K times `bytes_per_slot`) or restore-then-replay of `matched+1` positions. This is where the two K=3 workers died (`bench/freetoken-mtp-k1-spec.md:8-15`); the replay variant is the safer one because it reuses the width-1 kernels.

And the constraint from the K=1 merge (52d737a) still holds on the tiers: the verify routes 2 tokens through the CPU executor at roughly twice the distinct-expert cost, so the marginal token is not cheap here the way it is on a resident model (Unsloth reports 1.3 to 1.7x on one RTX 6000 PRO with the model resident). Expected on the 4090: graphs under MTP take K=1 from minus 33 percent to around break-even or a little better; K>1 gains diminish as distinct cold experts per step grow. Worth doing for the measurement and for the resident case, but after #5.

**Comment by jomcgi** [comment]
## Correction: graphs are necessary, probably not sufficient

I earlier characterised the blog's explanation of the MTP loss as wrong. It is not, and the arithmetic says the ceiling here is low even with a perfectly captured verify pass.

The post's claim is "speculation pays when a marginal token is cheap. With part of every step on the CPU tier a drafted token costs full price". That is exactly right for an offloaded MoE. Splitting the roughly 45 ms step by what a width-2 verify amortises:

- **Paid once per forward, so amortised:** dense and attention weight streaming, kernel launches, PLE gather. About 26 ms.
- **Scales with width, so not amortised:** MoE expert fetch and compute, about 18 ms, because two tokens route to roughly twice the distinct experts.

A width-2 verify is therefore about 2 x 18 + 26 = 62 ms against 90 ms for two separate steps. Cheaper, but the saving is only ever the fixed part, which is what the post was saying.

The rejection path then decides it. On `matched == 0` the engine restores state and replays a width-1 decode, so:

```
cost per token = (draft + verify + (1 - a) * replay) / (1 + a)
```

| acceptance | cost per token | against a 45 ms baseline |
|---|---|---|
| 52% | about 53 ms | worse |
| 65% | about 50 ms | worse |
| 78% | about 43 ms | about 4% better |

So break-even sits near the top of the measured 52 to 78 percent range. The missing CUDA graphs were a real and separable penalty, worth roughly the 40 percent gap between the 14 to 17 tok/s measured with MTP on and the 22 to 24 without, and this branch removes it. That gets MTP back to roughly parity rather than making it a win.

What would actually change the verdict, in order of expected value:

1. **K of 2 or more.** The fixed 26 ms amortises over more tokens, which is where the leverage is. Needs a mechanism for partial acceptance at K>1, which does not exist today.
2. **A cheaper rejection path.** A full width-1 replay after a miss is the single largest term at low acceptance. If the verify forward's seed position could be committed without re-running, the (1-a) term largely disappears.

Neither is a flag. Leaving this issue open with that as the plan rather than claiming the capture work closes it. The capture work is still worth having: it is a prerequisite for both, and three real capture bugs were fixed getting there (two constant indptr tensors built on the host, and an in-forward PLE prefetch that reached host I/O for staged backends).

**Comment by jomcgi** [comment]
## Measured: graphs work, MTP still loses. Closing as measured-negative.

The verify pass now captures and replays cleanly on the production uring PLE backend, zero crash markers, and the arithmetic in the previous comment held.

| arm | x1 tok/s (mean of 3) | decode batches (mean of 12) |
|---|---|---|
| MTP on, verify graph **off** | 16.1 | 17.5 |
| MTP on, verify graph **on** | **19.5** | **21.6** |
| MTP off entirely (production) | 21.5 | 25.7 |

Draft acceptance measured 767 accepted against 299 rejected, so **72.0 percent**, which matches the figure in the blog post.

So capturing the verify pass is worth about +21 percent over eager MTP, and that is real. It still leaves MTP about 9 percent behind not speculating at all. The prediction from the cost model was break-even near the top of the acceptance range, and 72 percent lands just under it. Nothing here is surprising any more, which is the point of having done the arithmetic first.

### Why it cannot be fixed with a flag

On this tier the drafted token does not get a discount. Roughly 26 ms of the step is per-forward and amortises across a width-2 verify (dense and attention weight streaming, launches, PLE gather), but roughly 18 ms is MoE expert fetch and compute, and that scales with width because the second token routes to its own experts. On top, a rejection replays a width-1 decode, so at 72 percent acceptance about 28 percent of steps pay an extra full step.

### What was fixed getting here, all worth keeping

Five bugs, four of them one class, host work or width-1 assumptions inside a captured region:

1. `ple.py` fused indptr built with `torch.tensor([0, 2], device=cuda)`, a host copy during capture.
2. `ple.py` `_short_conv` did the same with `.new_tensor([0, 1])`.
3. `model.py` called `ple.start_prefetch` inside the captured region, reaching host I/O for staged backends. Now guarded by an explicit per-step staging flag.
4. `engine/graph.py` sliced `active_table_idx` per request when route collection is per token.
5. `attention/qsa_sparse.py` staged the attention block table per request; the fused path now has its own token-sized table.

The tests parametrise over width so this class is caught without a GPU.

### Verdict

Closing as measured-negative on this box. The capture work stays: it is correct, it is a prerequisite for anything better, and it is the only reason we now have a real number instead of a suspicion.

Reopen if either of these lands, since both attack the actual cost terms:
- **K of 2 or more**, so the 26 ms fixed cost amortises further. Needs partial acceptance at K>1, which does not exist.
- **A cheaper rejection**, committing the verify pass's seed position without a full width-1 replay. That removes the largest term at this acceptance rate.

Also worth revisiting on a box where the model is resident, where the expert term does not scale with width and the usual speculative-decode economics apply. That is the GLM cloud case, not this one.

**Comment by jomcgi** [comment]
Agentic project reconciliation (2026-09-08): source and open PRs are ahead of the original issue.

Default branch `4fbc4eb` already has K=1 target verification in one model call, ordered width-one stateful operations, and snapshot/replay on rejection. Draft attention resets per draft call and ordinary graph capture is disabled when MTP is selected. Separately, fork #56-#63 implement target-cost/rollback/relocatable graph experiments; #63 reports width-five relocation correctness. #64 head `332c5737858839e59b72346a3593ab0870c90ec0` integrates causal ngram drafts, exact stop/rollback and optional paired CPU dots. It remains draft, one request only, and general serving improvement is unqualified: repetition improved, continuation results were mixed, and a small coding sample does not isolate engine throughput.

Reuse those implementations and measured outcomes before new K=2-4 work. Start by choosing the cheapest validated target-verification loop, then evaluate persistent MTP draft context, retained accepted state and bounded graph variants. Do not assume a fresh speculative implementation is needed.

Donor: [retained state](https://github.com/GenerelSchwerz/llama.cpp/wiki/Retained-MTP-State), [atomic verification](https://github.com/GenerelSchwerz/llama.cpp/wiki/Atomic-Speculative-Verification), [graph variants](https://github.com/GenerelSchwerz/llama.cpp/wiki/Speculative-MoE-Graph-Variants). Translate into FreeToken's QSA/GDN/PLE ownership; accepted-state retention reduces repeated work, while persistent context may separately improve proposal quality.

K=1 necessary cost test: draft + verify + expected snapshot/rollback overhead < (1 + acceptance rate) * ordinary decode cost. Measure it rather than projecting from acceptance alone. Exact rejection-prefix state, subsequent ordinary decode, EOS/stops/tool boundaries, page relocation and cancellation are required. Two-stream speculation remains a separate extension after ordinary two-stream qualification.

Project: https://github.com/orgs/jomcgi-org/projects/2
